spec/ngx_helper.lua

```diff
@@ -28,11 +30,24 @@ local set_status = getlocal(register_setter, 'ngx_magic_key_setters').status
 local get_headers_sent = getlocal(register_getter, 'ngx_magic_key_getters').headers_sent
 local set_headers_sent = getlocal(register_setter, 'ngx_magic_key_setters').headers_sent
 
+local function reset_ngx_shared(state)
+  local shared = ngx.shared
+
+  for name, dict in pairs(state) do
+    shared[name] = dict
+  end
+
+  for name, dict in pairs(shared) do
+    shared[name] = state[name]
+  end
+end
+
 local function reset_ngx_state()
   ngx.var = deepcopy(ngx_var_original)
   ngx.ctx = deepcopy(ngx_ctx_original)
-  ngx.shared = deepcopy(ngx_shared_original)
   ngx.header = deepcopy(ngx_header_original)
+
+  reset_ngx_shared(ngx_shared_original)
```

Comment:
Why doesn't `deepcopy()` work in this case?

Reply:
`deepcopy` works fine, but `resty.limit.req` takes a reference to `ngx.shared` when it is loaded:
https://github.com/openresty/lua-resty-limit-traffic/blob/ef080731b80fe00a385a4aed8a7f7f575d01c152/lib/resty/limit/req.lua#L11
So replacing the whole table won't have an effect.

Reply:
:+1:
I think this deserves a comment.